# **CLASIFICACIÓN NO CONVENCIONAL: MULTILABEL**
Máster Inteligencia Computacional e Internet de las Cosas

# **Apartado 1**

In [ ]:
!pip install wheel
!pip install scikit-multilearn #Successfully installed scikit-multilearn-0.2.0
!pip install liac-arff
!pip install scipy
!pip install requests
!pip install scikit-learn

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
     ━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━━ 89.4/89.4 KB 3.1 MB/s eta 0:00:00
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
  Preparing metadata (setup.py) ... done
  Created wheel for liac-arff: filename=liac_arff-2.5.0-py3-none-any.whl size=11732 sha256=feb47cd2bcf9ee8fcd07f0e7a3a9a9387cd3a4d446b0b49f1a193154e1663a0f
  Stored in directory: /root/.cache/pip/wheels/a2/de/68/bf3972de3ecb31e32bef59a7f4c75f0687a3674c476b347c14
Successfully built liac-arff
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/s

In [ ]:
#Realización del tutorial

In [ ]:
from skmultilearn.dataset import load_dataset

In [ ]:
X_train, y_train, feature_names, label_names = load_dataset('emotions', 'train')
X_test, y_test, _, _ = load_dataset('emotions', 'test')

emotions:train - does not exists downloading
Downloaded emotions-train
emotions:test - does not exists downloading
Downloaded emotions-test


In [ ]:
feature_names[:10]

[('Mean_Acc1298_Mean_Mem40_Centroid', 'NUMERIC'),
 ('Mean_Acc1298_Mean_Mem40_Rolloff', 'NUMERIC'),
 ('Mean_Acc1298_Mean_Mem40_Flux', 'NUMERIC'),
 ('Mean_Acc1298_Mean_Mem40_MFCC_0', 'NUMERIC'),
 ('Mean_Acc1298_Mean_Mem40_MFCC_1', 'NUMERIC'),
 ('Mean_Acc1298_Mean_Mem40_MFCC_2', 'NUMERIC'),
 ('Mean_Acc1298_Mean_Mem40_MFCC_3', 'NUMERIC'),
 ('Mean_Acc1298_Mean_Mem40_MFCC_4', 'NUMERIC'),
 ('Mean_Acc1298_Mean_Mem40_MFCC_5', 'NUMERIC'),
 ('Mean_Acc1298_Mean_Mem40_MFCC_6', 'NUMERIC')]

In [ ]:
label_names

[('amazed-suprised', ['0', '1']),
 ('happy-pleased', ['0', '1']),
 ('relaxing-calm', ['0', '1']),
 ('quiet-still', ['0', '1']),
 ('sad-lonely', ['0', '1']),
 ('angry-aggresive', ['0', '1'])]

In [ ]:
from skmultilearn.problem_transform import BinaryRelevance
from sklearn.svm import SVC

In [ ]:
clf = BinaryRelevance(
    classifier=SVC(),
    require_dense=[False, True]
)

In [ ]:
clf.classifiers_

In [ ]:
clf.fit(X_train, y_train)

BinaryRelevance(classifier=SVC(), require_dense=[False, True])

In [ ]:
clf.classifiers_

[SVC(), SVC(), SVC(), SVC(), SVC(), SVC()]

In [ ]:
prediction = clf.predict(X_test)

In [ ]:
prediction

<202x6 sparse matrix of type '<class 'numpy.int64'>'
	with 27 stored elements in Compressed Sparse Column format>

In [ ]:
import sklearn.metrics as metrics

In [ ]:
metrics.hamming_loss(y_test, prediction)

0.3250825082508251

In [ ]:
metrics.accuracy_score(y_test, prediction)

0.01485148514851485

In [ ]:
#Fin del tutorial, se realiza la carga y comprension de los 17 dataset

#**Apartado 2**

In [ ]:
#{'scene', 'Corel5k', 'bibtex', 'enron', 'rcv1subset5', 'tmc2007_500',
#'rcv1subset3', 'rcv1subset1', 'delicious', 'rcv1subset4', 'genbase', 'birds', 'emotions',
#'rcv1subset2', 'mediamill', 'medical', 'yeast'}

In [ ]:
#scene, dataset con distintos escenarios naturales como playa, urbano, montaña, etc

In [ ]:
X_train, y_train, feature_names, label_names = load_dataset('scene', 'train')
X_test, y_test, _, _ = load_dataset('scene', 'test')

scene:train - exists, not redownloading
scene:test - exists, not redownloading


In [ ]:
feature_names[:10]

[('Att1', 'NUMERIC'),
 ('Att2', 'NUMERIC'),
 ('Att3', 'NUMERIC'),
 ('Att4', 'NUMERIC'),
 ('Att5', 'NUMERIC'),
 ('Att6', 'NUMERIC'),
 ('Att7', 'NUMERIC'),
 ('Att8', 'NUMERIC'),
 ('Att9', 'NUMERIC'),
 ('Att10', 'NUMERIC')]

In [ ]:
label_names

[('Beach', ['0', '1']),
 ('Sunset', ['0', '1']),
 ('FallFoliage', ['0', '1']),
 ('Field', ['0', '1']),
 ('Mountain', ['0', '1']),
 ('Urban', ['0', '1'])]

In [ ]:
#Corel5k, dataset con muchisimas etiquetas, diferentes palabras desde animales, plantas, etc

In [ ]:
X_train, y_train, feature_names, label_names = load_dataset('Corel5k', 'train')
X_test, y_test, _, _ = load_dataset('Corel5k', 'test')

Corel5k:train - exists, not redownloading
Corel5k:test - exists, not redownloading


In [ ]:
feature_names[:10]

[('Cluster1', ['0', '1']),
 ('Cluster2', ['0', '1']),
 ('Cluster3', ['0', '1']),
 ('Cluster4', ['0', '1']),
 ('Cluster5', ['0', '1']),
 ('Cluster6', ['0', '1']),
 ('Cluster7', ['0', '1']),
 ('Cluster8', ['0', '1']),
 ('Cluster9', ['0', '1']),
 ('Cluster10', ['0', '1'])]

In [ ]:
label_names

[('city', ['0', '1']),
 ('mountain', ['0', '1']),
 ('sky', ['0', '1']),
 ('sun', ['0', '1']),
 ('water', ['0', '1']),
 ('clouds', ['0', '1']),
 ('tree', ['0', '1']),
 ('bay', ['0', '1']),
 ('lake', ['0', '1']),
 ('sea', ['0', '1']),
 ('beach', ['0', '1']),
 ('boats', ['0', '1']),
 ('people', ['0', '1']),
 ('branch', ['0', '1']),
 ('leaf', ['0', '1']),
 ('grass', ['0', '1']),
 ('plain', ['0', '1']),
 ('palm', ['0', '1']),
 ('horizon', ['0', '1']),
 ('shell', ['0', '1']),
 ('hills', ['0', '1']),
 ('waves', ['0', '1']),
 ('birds', ['0', '1']),
 ('land', ['0', '1']),
 ('dog', ['0', '1']),
 ('bridge', ['0', '1']),
 ('ships', ['0', '1']),
 ('buildings', ['0', '1']),
 ('fence', ['0', '1']),
 ('island', ['0', '1']),
 ('storm', ['0', '1']),
 ('peaks', ['0', '1']),
 ('jet', ['0', '1']),
 ('plane', ['0', '1']),
 ('runway', ['0', '1']),
 ('basket', ['0', '1']),
 ('flight', ['0', '1']),
 ('flag', ['0', '1']),
 ('helicopter', ['0', '1']),
 ('boeing', ['0', '1']),
 ('prop', ['0', '1']),
 ('f-16', ['0

In [ ]:
#bibtex, dataset diferentes labels sobre referencias topicos, web, simulaciones, estructuras, etc

In [ ]:
X_train, y_train, feature_names, label_names = load_dataset('bibtex', 'train')
X_test, y_test, _, _ = load_dataset('bibtex', 'test')

bibtex:train - exists, not redownloading
bibtex:test - exists, not redownloading


In [ ]:
feature_names[:10]

[('0', ['0', '1']),
 ('000', ['0', '1']),
 ('02', ['0', '1']),
 ('05', ['0', '1']),
 ('06', ['0', '1']),
 ('1', ['0', '1']),
 ('10', ['0', '1']),
 ('100', ['0', '1']),
 ('11', ['0', '1']),
 ('12', ['0', '1'])]

In [ ]:
label_names

[('TAG_2005', ['0', '1']),
 ('TAG_2006', ['0', '1']),
 ('TAG_2007', ['0', '1']),
 ('TAG_agdetection', ['0', '1']),
 ('TAG_algorithms', ['0', '1']),
 ('TAG_amperometry', ['0', '1']),
 ('TAG_analysis', ['0', '1']),
 ('TAG_and', ['0', '1']),
 ('TAG_annotation', ['0', '1']),
 ('TAG_antibody', ['0', '1']),
 ('TAG_apob', ['0', '1']),
 ('TAG_architecture', ['0', '1']),
 ('TAG_article', ['0', '1']),
 ('TAG_bettasplendens', ['0', '1']),
 ('TAG_bibteximport', ['0', '1']),
 ('TAG_book', ['0', '1']),
 ('TAG_children', ['0', '1']),
 ('TAG_classification', ['0', '1']),
 ('TAG_clustering', ['0', '1']),
 ('TAG_cognition', ['0', '1']),
 ('TAG_collaboration', ['0', '1']),
 ('TAG_collaborative', ['0', '1']),
 ('TAG_community', ['0', '1']),
 ('TAG_competition', ['0', '1']),
 ('TAG_complex', ['0', '1']),
 ('TAG_complexity', ['0', '1']),
 ('TAG_compounds', ['0', '1']),
 ('TAG_computer', ['0', '1']),
 ('TAG_computing', ['0', '1']),
 ('TAG_concept', ['0', '1']),
 ('TAG_context', ['0', '1']),
 ('TAG_cortex', [

In [ ]:
#enron, dataset diferentes labels sobre codificaciones como A.A8, C.C9, D.D1

In [ ]:
X_train, y_train, feature_names, label_names = load_dataset('enron', 'train')
X_test, y_test, _, _ = load_dataset('enron', 'test')

enron:train - exists, not redownloading
enron:test - exists, not redownloading


In [ ]:
feature_names[:10]

[('0', ['0', '1']),
 ('00', ['0', '1']),
 ('000', ['0', '1']),
 ('01', ['0', '1']),
 ('02', ['0', '1']),
 ('03', ['0', '1']),
 ('04', ['0', '1']),
 ('05', ['0', '1']),
 ('06', ['0', '1']),
 ('07', ['0', '1'])]

In [ ]:
label_names

[('A.A8', ['0', '1']),
 ('C.C9', ['0', '1']),
 ('B.B12', ['0', '1']),
 ('C.C11', ['0', '1']),
 ('C.C5', ['0', '1']),
 ('C.C7', ['0', '1']),
 ('B.B2', ['0', '1']),
 ('B.B3', ['0', '1']),
 ('D.D16', ['0', '1']),
 ('A.A7', ['0', '1']),
 ('D.D1', ['0', '1']),
 ('A.A4', ['0', '1']),
 ('C.C2', ['0', '1']),
 ('A.A3', ['0', '1']),
 ('A.A1', ['0', '1']),
 ('D.D9', ['0', '1']),
 ('D.D19', ['0', '1']),
 ('B.B8', ['0', '1']),
 ('D.D12', ['0', '1']),
 ('D.D6', ['0', '1']),
 ('C.C8', ['0', '1']),
 ('A.A6', ['0', '1']),
 ('B.B9', ['0', '1']),
 ('A.A5', ['0', '1']),
 ('C.C10', ['0', '1']),
 ('B.B1', ['0', '1']),
 ('D.D5', ['0', '1']),
 ('B.B11', ['0', '1']),
 ('D.D2', ['0', '1']),
 ('B.B4', ['0', '1']),
 ('D.D15', ['0', '1']),
 ('C.C4', ['0', '1']),
 ('D.D8', ['0', '1']),
 ('B.B6', ['0', '1']),
 ('D.D3', ['0', '1']),
 ('D.D13', ['0', '1']),
 ('D.D7', ['0', '1']),
 ('C.C12', ['0', '1']),
 ('B.B7', ['0', '1']),
 ('C.C6', ['0', '1']),
 ('B.B5', ['0', '1']),
 ('D.D11', ['0', '1']),
 ('A.A2', ['0', '1']),


In [ ]:
#rcv1subset5, dataset con diferentes labels desde clase 1 hasta clase 101

In [ ]:
X_train, y_train, feature_names, label_names = load_dataset('rcv1subset5', 'train')
X_test, y_test, _, _ = load_dataset('rcv1subset5', 'test')

rcv1subset5:train - exists, not redownloading
rcv1subset5:test - exists, not redownloading


In [ ]:
feature_names[:10]

[('Att1', 'NUMERIC'),
 ('Att2', 'NUMERIC'),
 ('Att3', 'NUMERIC'),
 ('Att4', 'NUMERIC'),
 ('Att5', 'NUMERIC'),
 ('Att6', 'NUMERIC'),
 ('Att7', 'NUMERIC'),
 ('Att8', 'NUMERIC'),
 ('Att9', 'NUMERIC'),
 ('Att10', 'NUMERIC')]

In [ ]:
label_names

[('Class1', ['0', '1']),
 ('Class2', ['0', '1']),
 ('Class3', ['0', '1']),
 ('Class4', ['0', '1']),
 ('Class5', ['0', '1']),
 ('Class6', ['0', '1']),
 ('Class7', ['0', '1']),
 ('Class8', ['0', '1']),
 ('Class9', ['0', '1']),
 ('Class10', ['0', '1']),
 ('Class11', ['0', '1']),
 ('Class12', ['0', '1']),
 ('Class13', ['0', '1']),
 ('Class14', ['0', '1']),
 ('Class15', ['0', '1']),
 ('Class16', ['0', '1']),
 ('Class17', ['0', '1']),
 ('Class18', ['0', '1']),
 ('Class19', ['0', '1']),
 ('Class20', ['0', '1']),
 ('Class21', ['0', '1']),
 ('Class22', ['0', '1']),
 ('Class23', ['0', '1']),
 ('Class24', ['0', '1']),
 ('Class25', ['0', '1']),
 ('Class26', ['0', '1']),
 ('Class27', ['0', '1']),
 ('Class28', ['0', '1']),
 ('Class29', ['0', '1']),
 ('Class30', ['0', '1']),
 ('Class31', ['0', '1']),
 ('Class32', ['0', '1']),
 ('Class33', ['0', '1']),
 ('Class34', ['0', '1']),
 ('Class35', ['0', '1']),
 ('Class36', ['0', '1']),
 ('Class37', ['0', '1']),
 ('Class38', ['0', '1']),
 ('Class39', ['0', '1

In [ ]:
#tmc2007_500, dataset con diferentes labels desde clase 1 hasta clase 22

In [ ]:
X_train, y_train, feature_names, label_names = load_dataset('tmc2007_500', 'train')
X_test, y_test, _, _ = load_dataset('tmc2007_500', 'test')

tmc2007_500:train - exists, not redownloading
tmc2007_500:test - exists, not redownloading


In [ ]:
feature_names[:10]

[('abort', ['0', '1']),
 ('abortedtakeoff', ['0', '1']),
 ('above', ['0', '1']),
 ('abuse', ['0', '1']),
 ('abusive', ['0', '1']),
 ('accomplish', ['0', '1']),
 ('acquire', ['0', '1']),
 ('acrid', ['0', '1']),
 ('across', ['0', '1']),
 ('activerunway', ['0', '1'])]

In [ ]:
label_names

[('class01', ['0', '1']),
 ('class02', ['0', '1']),
 ('class03', ['0', '1']),
 ('class04', ['0', '1']),
 ('class05', ['0', '1']),
 ('class06', ['0', '1']),
 ('class07', ['0', '1']),
 ('class08', ['0', '1']),
 ('class09', ['0', '1']),
 ('class10', ['0', '1']),
 ('class11', ['0', '1']),
 ('class12', ['0', '1']),
 ('class13', ['0', '1']),
 ('class14', ['0', '1']),
 ('class15', ['0', '1']),
 ('class16', ['0', '1']),
 ('class17', ['0', '1']),
 ('class18', ['0', '1']),
 ('class19', ['0', '1']),
 ('class20', ['0', '1']),
 ('class21', ['0', '1']),
 ('class22', ['0', '1'])]

In [ ]:
#rcv1subset3, dataset con diferentes labels desde clase 1 hasta clase 101, similar a rcv1subset5 pero usando otros subconjunto

In [ ]:
X_train, y_train, feature_names, label_names = load_dataset('rcv1subset3', 'train')
X_test, y_test, _, _ = load_dataset('rcv1subset3', 'test')

rcv1subset3:train - exists, not redownloading
rcv1subset3:test - exists, not redownloading


In [ ]:
feature_names[:10]

[('Att1', 'NUMERIC'),
 ('Att2', 'NUMERIC'),
 ('Att3', 'NUMERIC'),
 ('Att4', 'NUMERIC'),
 ('Att5', 'NUMERIC'),
 ('Att6', 'NUMERIC'),
 ('Att7', 'NUMERIC'),
 ('Att8', 'NUMERIC'),
 ('Att9', 'NUMERIC'),
 ('Att10', 'NUMERIC')]

In [ ]:
label_names

[('Class1', ['0', '1']),
 ('Class2', ['0', '1']),
 ('Class3', ['0', '1']),
 ('Class4', ['0', '1']),
 ('Class5', ['0', '1']),
 ('Class6', ['0', '1']),
 ('Class7', ['0', '1']),
 ('Class8', ['0', '1']),
 ('Class9', ['0', '1']),
 ('Class10', ['0', '1']),
 ('Class11', ['0', '1']),
 ('Class12', ['0', '1']),
 ('Class13', ['0', '1']),
 ('Class14', ['0', '1']),
 ('Class15', ['0', '1']),
 ('Class16', ['0', '1']),
 ('Class17', ['0', '1']),
 ('Class18', ['0', '1']),
 ('Class19', ['0', '1']),
 ('Class20', ['0', '1']),
 ('Class21', ['0', '1']),
 ('Class22', ['0', '1']),
 ('Class23', ['0', '1']),
 ('Class24', ['0', '1']),
 ('Class25', ['0', '1']),
 ('Class26', ['0', '1']),
 ('Class27', ['0', '1']),
 ('Class28', ['0', '1']),
 ('Class29', ['0', '1']),
 ('Class30', ['0', '1']),
 ('Class31', ['0', '1']),
 ('Class32', ['0', '1']),
 ('Class33', ['0', '1']),
 ('Class34', ['0', '1']),
 ('Class35', ['0', '1']),
 ('Class36', ['0', '1']),
 ('Class37', ['0', '1']),
 ('Class38', ['0', '1']),
 ('Class39', ['0', '1

In [ ]:
#rcv1subset1, dataset con diferentes labels desde clase 1 hasta clase 101, similar a rcv1subset5 pero usando otros subconjunto

In [ ]:
X_train, y_train, feature_names, label_names = load_dataset('rcv1subset1', 'train')
X_test, y_test, _, _ = load_dataset('rcv1subset1', 'test')

rcv1subset1:train - exists, not redownloading
rcv1subset1:test - exists, not redownloading


In [ ]:
feature_names[:10]

[('Att1', 'NUMERIC'),
 ('Att2', 'NUMERIC'),
 ('Att3', 'NUMERIC'),
 ('Att4', 'NUMERIC'),
 ('Att5', 'NUMERIC'),
 ('Att6', 'NUMERIC'),
 ('Att7', 'NUMERIC'),
 ('Att8', 'NUMERIC'),
 ('Att9', 'NUMERIC'),
 ('Att10', 'NUMERIC')]

In [ ]:
label_names

[('Class1', ['0', '1']),
 ('Class2', ['0', '1']),
 ('Class3', ['0', '1']),
 ('Class4', ['0', '1']),
 ('Class5', ['0', '1']),
 ('Class6', ['0', '1']),
 ('Class7', ['0', '1']),
 ('Class8', ['0', '1']),
 ('Class9', ['0', '1']),
 ('Class10', ['0', '1']),
 ('Class11', ['0', '1']),
 ('Class12', ['0', '1']),
 ('Class13', ['0', '1']),
 ('Class14', ['0', '1']),
 ('Class15', ['0', '1']),
 ('Class16', ['0', '1']),
 ('Class17', ['0', '1']),
 ('Class18', ['0', '1']),
 ('Class19', ['0', '1']),
 ('Class20', ['0', '1']),
 ('Class21', ['0', '1']),
 ('Class22', ['0', '1']),
 ('Class23', ['0', '1']),
 ('Class24', ['0', '1']),
 ('Class25', ['0', '1']),
 ('Class26', ['0', '1']),
 ('Class27', ['0', '1']),
 ('Class28', ['0', '1']),
 ('Class29', ['0', '1']),
 ('Class30', ['0', '1']),
 ('Class31', ['0', '1']),
 ('Class32', ['0', '1']),
 ('Class33', ['0', '1']),
 ('Class34', ['0', '1']),
 ('Class35', ['0', '1']),
 ('Class36', ['0', '1']),
 ('Class37', ['0', '1']),
 ('Class38', ['0', '1']),
 ('Class39', ['0', '1

In [ ]:
#rcv1subset4, dataset con diferentes labels desde clase 1 hasta clase 101, similar a rcv1subset5 pero usando otros subconjunto

In [ ]:
X_train, y_train, feature_names, label_names = load_dataset('rcv1subset4', 'train')
X_test, y_test, _, _ = load_dataset('rcv1subset4', 'test')

rcv1subset4:train - exists, not redownloading
rcv1subset4:test - exists, not redownloading


In [ ]:
feature_names[:10]

[('Att1', 'NUMERIC'),
 ('Att2', 'NUMERIC'),
 ('Att3', 'NUMERIC'),
 ('Att4', 'NUMERIC'),
 ('Att5', 'NUMERIC'),
 ('Att6', 'NUMERIC'),
 ('Att7', 'NUMERIC'),
 ('Att8', 'NUMERIC'),
 ('Att9', 'NUMERIC'),
 ('Att10', 'NUMERIC')]

In [ ]:
label_names

[('Class1', ['0', '1']),
 ('Class2', ['0', '1']),
 ('Class3', ['0', '1']),
 ('Class4', ['0', '1']),
 ('Class5', ['0', '1']),
 ('Class6', ['0', '1']),
 ('Class7', ['0', '1']),
 ('Class8', ['0', '1']),
 ('Class9', ['0', '1']),
 ('Class10', ['0', '1']),
 ('Class11', ['0', '1']),
 ('Class12', ['0', '1']),
 ('Class13', ['0', '1']),
 ('Class14', ['0', '1']),
 ('Class15', ['0', '1']),
 ('Class16', ['0', '1']),
 ('Class17', ['0', '1']),
 ('Class18', ['0', '1']),
 ('Class19', ['0', '1']),
 ('Class20', ['0', '1']),
 ('Class21', ['0', '1']),
 ('Class22', ['0', '1']),
 ('Class23', ['0', '1']),
 ('Class24', ['0', '1']),
 ('Class25', ['0', '1']),
 ('Class26', ['0', '1']),
 ('Class27', ['0', '1']),
 ('Class28', ['0', '1']),
 ('Class29', ['0', '1']),
 ('Class30', ['0', '1']),
 ('Class31', ['0', '1']),
 ('Class32', ['0', '1']),
 ('Class33', ['0', '1']),
 ('Class34', ['0', '1']),
 ('Class35', ['0', '1']),
 ('Class36', ['0', '1']),
 ('Class37', ['0', '1']),
 ('Class38', ['0', '1']),
 ('Class39', ['0', '1

In [ ]:
#rcv1subset2, dataset con diferentes labels desde clase 1 hasta clase 101, similar a rcv1subset5 pero usando otros subconjunto

In [ ]:
X_train, y_train, feature_names, label_names = load_dataset('rcv1subset2', 'train')
X_test, y_test, _, _ = load_dataset('rcv1subset2', 'test')

rcv1subset2:train - exists, not redownloading
rcv1subset2:test - exists, not redownloading


In [ ]:
feature_names[:10]

[('Att1', 'NUMERIC'),
 ('Att2', 'NUMERIC'),
 ('Att3', 'NUMERIC'),
 ('Att4', 'NUMERIC'),
 ('Att5', 'NUMERIC'),
 ('Att6', 'NUMERIC'),
 ('Att7', 'NUMERIC'),
 ('Att8', 'NUMERIC'),
 ('Att9', 'NUMERIC'),
 ('Att10', 'NUMERIC')]

In [ ]:
label_names

[('Class1', ['0', '1']),
 ('Class2', ['0', '1']),
 ('Class3', ['0', '1']),
 ('Class4', ['0', '1']),
 ('Class5', ['0', '1']),
 ('Class6', ['0', '1']),
 ('Class7', ['0', '1']),
 ('Class8', ['0', '1']),
 ('Class9', ['0', '1']),
 ('Class10', ['0', '1']),
 ('Class11', ['0', '1']),
 ('Class12', ['0', '1']),
 ('Class13', ['0', '1']),
 ('Class14', ['0', '1']),
 ('Class15', ['0', '1']),
 ('Class16', ['0', '1']),
 ('Class17', ['0', '1']),
 ('Class18', ['0', '1']),
 ('Class19', ['0', '1']),
 ('Class20', ['0', '1']),
 ('Class21', ['0', '1']),
 ('Class22', ['0', '1']),
 ('Class23', ['0', '1']),
 ('Class24', ['0', '1']),
 ('Class25', ['0', '1']),
 ('Class26', ['0', '1']),
 ('Class27', ['0', '1']),
 ('Class28', ['0', '1']),
 ('Class29', ['0', '1']),
 ('Class30', ['0', '1']),
 ('Class31', ['0', '1']),
 ('Class32', ['0', '1']),
 ('Class33', ['0', '1']),
 ('Class34', ['0', '1']),
 ('Class35', ['0', '1']),
 ('Class36', ['0', '1']),
 ('Class37', ['0', '1']),
 ('Class38', ['0', '1']),
 ('Class39', ['0', '1

In [ ]:
#delicious, dataset con diferentes labels con diferentes Tag como linux, web, logo, etc

In [ ]:
X_train, y_train, feature_names, label_names = load_dataset('delicious', 'train')
X_test, y_test, _, _ = load_dataset('delicious', 'test')

delicious:train - exists, not redownloading
delicious:test - exists, not redownloading


In [ ]:
feature_names[:10]

[('_qacct', ['0', '1']),
 ('accessing', ['0', '1']),
 ('actionscript', ['0', '1']),
 ('activerecord', ['0', '1']),
 ('addoverlay', ['0', '1']),
 ('afternoon', ['0', '1']),
 ('against', ['0', '1']),
 ('air', ['0', '1']),
 ('ajax_action', ['0', '1']),
 ('als', ['0', '1'])]

In [ ]:
label_names

[('TAG_.imported', ['0', '1']),
 ('TAG_.net', ['0', '1']),
 ('TAG_2.0', ['0', '1']),
 ('TAG_2007', ['0', '1']),
 ('TAG_3d', ['0', '1']),
 ('TAG_??', ['0', '1']),
 ('TAG_???', ['0', '1']),
 ('TAG_????', ['0', '1']),
 ('TAG_academia', ['0', '1']),
 ('TAG_academic', ['0', '1']),
 ('TAG_access', ['0', '1']),
 ('TAG_accessibility', ['0', '1']),
 ('TAG_accessories', ['0', '1']),
 ('TAG_accounts', ['0', '1']),
 ('TAG_actionscript', ['0', '1']),
 ('TAG_activism', ['0', '1']),
 ('TAG_ad', ['0', '1']),
 ('TAG_addon', ['0', '1']),
 ('TAG_addons', ['0', '1']),
 ('TAG_admin', ['0', '1']),
 ('TAG_administration', ['0', '1']),
 ('TAG_adobe', ['0', '1']),
 ('TAG_ads', ['0', '1']),
 ('TAG_adsense', ['0', '1']),
 ('TAG_adult', ['0', '1']),
 ('TAG_advertising', ['0', '1']),
 ('TAG_advice', ['0', '1']),
 ('TAG_affiliate', ['0', '1']),
 ('TAG_agencies', ['0', '1']),
 ('TAG_agency', ['0', '1']),
 ('TAG_aggregator', ['0', '1']),
 ('TAG_agile', ['0', '1']),
 ('TAG_ai', ['0', '1']),
 ('TAG_air', ['0', '1']),
 

In [ ]:
#genbase, dataset con diferentes labels, que contiene secuencias de proteinas que pueden ser asignadas a otras proteinas familiares
#un ejemplo de label puede ser PDOC00154 o PDOC00343

In [ ]:
X_train, y_train, feature_names, label_names = load_dataset('genbase', 'train')
X_test, y_test, _, _ = load_dataset('genbase', 'test')

genbase:train - exists, not redownloading
genbase:test - exists, not redownloading


In [ ]:
feature_names[:10]

[('protein',
  ['O00060',
   'O00139',
   'O02741',
   'O08424',
   'O12984',
   'O12988',
   'O13224',
   'O13351',
   'O13433',
   'O14343',
   'O14558',
   'O14782',
   'O14788',
   'O14936',
   'O15066',
   'O15886',
   'O19131',
   'O23826',
   'O24585',
   'O24594',
   'O26662',
   'O28538',
   'O30851',
   'O31288',
   'O34328',
   'O35066',
   'O35071',
   'O35235',
   'O35734',
   'O35787',
   'O35878',
   'O43093',
   'O43416',
   'O43557',
   'O43896',
   'O49886',
   'O49939',
   'O51628',
   'O53021',
   'O55165',
   'O59469',
   'O60282',
   'O60333',
   'O62683',
   'O64966',
   'O64967',
   'O66105',
   'O69241',
   'O69242',
   'O70589',
   'O73919',
   'O74164',
   'O75001',
   'O75888',
   'O76819',
   'O77510',
   'O77736',
   'O77764',
   'O84033',
   'O86110',
   'O88658',
   'O93591',
   'O95049',
   'O95235',
   'O95239',
   'O97605',
   'O97626',
   'P00347',
   'P01374',
   'P01375',
   'P02248',
   'P02470',
   'P02471',
   'P02472',
   'P02473',
   'P02474',

In [ ]:
label_names

[('PDOC00154', ['0', '1']),
 ('PDOC00343', ['0', '1']),
 ('PDOC00271', ['0', '1']),
 ('PDOC00064', ['0', '1']),
 ('PDOC00791', ['0', '1']),
 ('PDOC00380', ['0', '1']),
 ('PDOC50007', ['0', '1']),
 ('PDOC00224', ['0', '1']),
 ('PDOC00100', ['0', '1']),
 ('PDOC00670', ['0', '1']),
 ('PDOC50002', ['0', '1']),
 ('PDOC50106', ['0', '1']),
 ('PDOC00561', ['0', '1']),
 ('PDOC50017', ['0', '1']),
 ('PDOC50003', ['0', '1']),
 ('PDOC50006', ['0', '1']),
 ('PDOC50156', ['0', '1']),
 ('PDOC00662', ['0', '1']),
 ('PDOC00018', ['0', '1']),
 ('PDOC50001', ['0', '1']),
 ('PDOC00014', ['0', '1']),
 ('PDOC00750', ['0', '1']),
 ('PDOC50196', ['0', '1']),
 ('PDOC50199', ['0', '1']),
 ('PDOC00660', ['0', '1']),
 ('PDOC00653', ['0', '1']),
 ('PDOC00030', ['0', '1'])]

In [ ]:
#birds, dataset con diferentes labels siendo los diferentes tipos de pájaros posibles. Dependerá de los distintos tipos de sonidos

In [ ]:
X_train, y_train, feature_names, label_names = load_dataset('birds', 'train')
X_test, y_test, _, _ = load_dataset('birds', 'test')

birds:train - exists, not redownloading
birds:test - exists, not redownloading


In [ ]:
feature_names[:10]

[('audio-ssd1', 'NUMERIC'),
 ('audio-ssd2', 'NUMERIC'),
 ('audio-ssd3', 'NUMERIC'),
 ('audio-ssd4', 'NUMERIC'),
 ('audio-ssd5', 'NUMERIC'),
 ('audio-ssd6', 'NUMERIC'),
 ('audio-ssd7', 'NUMERIC'),
 ('audio-ssd8', 'NUMERIC'),
 ('audio-ssd9', 'NUMERIC'),
 ('audio-ssd10', 'NUMERIC')]

In [ ]:
label_names

[('Brown Creeper', ['0', '1']),
 ('Pacific Wren', ['0', '1']),
 ('Pacific-slope Flycatcher', ['0', '1']),
 ('Red-breasted Nuthatch', ['0', '1']),
 ('Dark-eyed Junco', ['0', '1']),
 ('Olive-sided Flycatcher', ['0', '1']),
 ('Hermit Thrush', ['0', '1']),
 ('Chestnut-backed Chickadee', ['0', '1']),
 ('Varied Thrush', ['0', '1']),
 ('Hermit Warbler', ['0', '1']),
 ("Swainson\\'s Thrush", ['0', '1']),
 ("Hammond\\'s Flycatcher", ['0', '1']),
 ('Western Tanager', ['0', '1']),
 ('Black-headed Grosbeak', ['0', '1']),
 ('Golden Crowned Kinglet', ['0', '1']),
 ('Warbling Vireo', ['0', '1']),
 ("MacGillivray\\'s Warbler", ['0', '1']),
 ("Stellar\\'s Jay", ['0', '1']),
 ('Common Nighthawk', ['0', '1'])]

In [ ]:
#emotions, dataset con diferentes labels siendo los diferentes tippos de emociones posibles como amazed-suprised, relaxing-calm, etc,
#esta relacionado con la musica y las emociones, es decir, los inputs son canciones.

In [ ]:
X_train, y_train, feature_names, label_names = load_dataset('emotions', 'train')
X_test, y_test, _, _ = load_dataset('emotions', 'test')

emotions:train - exists, not redownloading
emotions:test - exists, not redownloading


In [ ]:
feature_names[:10]

[('Mean_Acc1298_Mean_Mem40_Centroid', 'NUMERIC'),
 ('Mean_Acc1298_Mean_Mem40_Rolloff', 'NUMERIC'),
 ('Mean_Acc1298_Mean_Mem40_Flux', 'NUMERIC'),
 ('Mean_Acc1298_Mean_Mem40_MFCC_0', 'NUMERIC'),
 ('Mean_Acc1298_Mean_Mem40_MFCC_1', 'NUMERIC'),
 ('Mean_Acc1298_Mean_Mem40_MFCC_2', 'NUMERIC'),
 ('Mean_Acc1298_Mean_Mem40_MFCC_3', 'NUMERIC'),
 ('Mean_Acc1298_Mean_Mem40_MFCC_4', 'NUMERIC'),
 ('Mean_Acc1298_Mean_Mem40_MFCC_5', 'NUMERIC'),
 ('Mean_Acc1298_Mean_Mem40_MFCC_6', 'NUMERIC')]

In [ ]:
label_names

[('amazed-suprised', ['0', '1']),
 ('happy-pleased', ['0', '1']),
 ('relaxing-calm', ['0', '1']),
 ('quiet-still', ['0', '1']),
 ('sad-lonely', ['0', '1']),
 ('angry-aggresive', ['0', '1'])]

In [ ]:
#mediamill, dataset con diferentes labels desde la clase 1 hasta la clase 101

In [ ]:
X_train, y_train, feature_names, label_names = load_dataset('mediamill', 'train')
X_test, y_test, _, _ = load_dataset('mediamill', 'test')

mediamill:train - exists, not redownloading
mediamill:test - exists, not redownloading


In [ ]:
feature_names[:10]

[('Att1', 'NUMERIC'),
 ('Att2', 'NUMERIC'),
 ('Att3', 'NUMERIC'),
 ('Att4', 'NUMERIC'),
 ('Att5', 'NUMERIC'),
 ('Att6', 'NUMERIC'),
 ('Att7', 'NUMERIC'),
 ('Att8', 'NUMERIC'),
 ('Att9', 'NUMERIC'),
 ('Att10', 'NUMERIC')]

In [ ]:
label_names

[('Class1', ['0', '1']),
 ('Class2', ['0', '1']),
 ('Class3', ['0', '1']),
 ('Class4', ['0', '1']),
 ('Class5', ['0', '1']),
 ('Class6', ['0', '1']),
 ('Class7', ['0', '1']),
 ('Class8', ['0', '1']),
 ('Class9', ['0', '1']),
 ('Class10', ['0', '1']),
 ('Class11', ['0', '1']),
 ('Class12', ['0', '1']),
 ('Class13', ['0', '1']),
 ('Class14', ['0', '1']),
 ('Class15', ['0', '1']),
 ('Class16', ['0', '1']),
 ('Class17', ['0', '1']),
 ('Class18', ['0', '1']),
 ('Class19', ['0', '1']),
 ('Class20', ['0', '1']),
 ('Class21', ['0', '1']),
 ('Class22', ['0', '1']),
 ('Class23', ['0', '1']),
 ('Class24', ['0', '1']),
 ('Class25', ['0', '1']),
 ('Class26', ['0', '1']),
 ('Class27', ['0', '1']),
 ('Class28', ['0', '1']),
 ('Class29', ['0', '1']),
 ('Class30', ['0', '1']),
 ('Class31', ['0', '1']),
 ('Class32', ['0', '1']),
 ('Class33', ['0', '1']),
 ('Class34', ['0', '1']),
 ('Class35', ['0', '1']),
 ('Class36', ['0', '1']),
 ('Class37', ['0', '1']),
 ('Class38', ['0', '1']),
 ('Class39', ['0', '1

In [ ]:
#medical, dataset con diferentes labels desde la clase 1 hasta la clase 44

In [ ]:
X_train, y_train, feature_names, label_names = load_dataset('medical', 'train')
X_test, y_test, _, _ = load_dataset('medical', 'test')

medical:train - exists, not redownloading
medical:test - exists, not redownloading


In [ ]:
feature_names[:10]

[('-', ['0', '1']),
 ('/', ['0', '1']),
 ('0', ['0', '1']),
 ('00', ['0', '1']),
 ('04', ['0', '1']),
 ('0;', ['0', '1']),
 ('0cm', ['0', '1']),
 ('1', ['0', '1']),
 ('1-1/2', ['0', '1']),
 ('1-1/2-year', ['0', '1'])]

In [ ]:
label_names

[('Class-0-593_70', ['0', '1']),
 ('Class-1-079_99', ['0', '1']),
 ('Class-2-786_09', ['0', '1']),
 ('Class-3-759_89', ['0', '1']),
 ('Class-4-753_0', ['0', '1']),
 ('Class-5-786_2', ['0', '1']),
 ('Class-6-V72_5', ['0', '1']),
 ('Class-7-511_9', ['0', '1']),
 ('Class-8-596_8', ['0', '1']),
 ('Class-9-599_0', ['0', '1']),
 ('Class-10-518_0', ['0', '1']),
 ('Class-11-593_5', ['0', '1']),
 ('Class-12-V13_09', ['0', '1']),
 ('Class-13-791_0', ['0', '1']),
 ('Class-14-789_00', ['0', '1']),
 ('Class-15-593_1', ['0', '1']),
 ('Class-16-462', ['0', '1']),
 ('Class-17-592_0', ['0', '1']),
 ('Class-18-786_59', ['0', '1']),
 ('Class-19-785_6', ['0', '1']),
 ('Class-20-V67_09', ['0', '1']),
 ('Class-21-795_5', ['0', '1']),
 ('Class-22-789_09', ['0', '1']),
 ('Class-23-786_50', ['0', '1']),
 ('Class-24-596_54', ['0', '1']),
 ('Class-25-787_03', ['0', '1']),
 ('Class-26-V42_0', ['0', '1']),
 ('Class-27-786_05', ['0', '1']),
 ('Class-28-753_21', ['0', '1']),
 ('Class-29-783_0', ['0', '1']),
 ('Class

In [ ]:
#yeast, dataset con diferentes labels desde la clase 1 hasta la clase 14, esta relacionado con la bioinformática

In [ ]:
X_train, y_train, feature_names, label_names = load_dataset('yeast', 'train')
X_test, y_test, _, _ = load_dataset('yeast', 'test')

yeast:train - exists, not redownloading
yeast:test - exists, not redownloading


In [ ]:
feature_names[:10]

[('Att1', 'NUMERIC'),
 ('Att2', 'NUMERIC'),
 ('Att3', 'NUMERIC'),
 ('Att4', 'NUMERIC'),
 ('Att5', 'NUMERIC'),
 ('Att6', 'NUMERIC'),
 ('Att7', 'NUMERIC'),
 ('Att8', 'NUMERIC'),
 ('Att9', 'NUMERIC'),
 ('Att10', 'NUMERIC')]

In [ ]:
label_names

[('Class1', ['0', '1']),
 ('Class2', ['0', '1']),
 ('Class3', ['0', '1']),
 ('Class4', ['0', '1']),
 ('Class5', ['0', '1']),
 ('Class6', ['0', '1']),
 ('Class7', ['0', '1']),
 ('Class8', ['0', '1']),
 ('Class9', ['0', '1']),
 ('Class10', ['0', '1']),
 ('Class11', ['0', '1']),
 ('Class12', ['0', '1']),
 ('Class13', ['0', '1']),
 ('Class14', ['0', '1'])]

#**Apartado 3**


In [ ]:
#car.py

In [ ]:
!pip install wheel
!pip install scikit-multilearn #Successfully installed scikit-multilearn-0.2.0
!pip install liac-arff
!pip install scipy
!pip install requests
!pip install scikit-learn

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [ ]:
#import necesarios

In [ ]:
from skmultilearn.dataset import load_dataset

In [ ]:
#codigo para conseguir los estadísticos de todos los dataset

In [ ]:

list = ['scene', 'Corel5k', 'bibtex', 'enron', 'rcv1subset5', 'tmc2007_500','rcv1subset3', 'rcv1subset1', 'delicious']

for i in list:
  X_train, y_train, feature_names, label_names = load_dataset(i, 'train')
  
  print(("number of instances -> " + str(X_train.shape[0])))#instancias
  
  print(("number of attributes -> " + str(X_train.shape[1])))#atributos
  
  print(("number of labels -> " + str(y_train.shape[1])))#etiquetas
  
  print("cardinality -> "+ str(y_train.sum()/X_train.shape[0]))#cardinalidad=promedio de etiquetas asociadas a cada instancia (total de 1/instancias)
  
  print("density -> "+ str((y_train.sum()/X_train.shape[0])/y_train.shape[1]))#densidad=cardinalidad entre el número total de etiquetas
  
  #codigo para sacar todas las combinaciones que tenemos y la frecuencia de cada combinacion
  uniquecomb = [y_train.getrow(0),]
  freqcomb = [1,]
  contador_combinaciones = 1
  for w in range(y_train.shape[0]):
    combunica = True
    for z in range(len(uniquecomb)):
      if (y_train.getrow(w).toarray() == uniquecomb[z]).all():
        combunica = False
        freqcomb[z] += 1
        break
    if combunica:
      uniquecomb.append(y_train.getrow(w))
      freqcomb.append(1)
      contador_combinaciones += 1 
  #for w in uniquecomb:
    #print(w.toarray()) 
  #print(contador)
  n_posibles_salidas= 2**y_train.shape[1]
  print("diversity -> "+ str(contador_combinaciones/n_posibles_salidas))#diversidad=medida Distinct(número de combinaciones de distintas etiquetas presentes en un conjunto de datos) 
  
  #normalizada por el número total de ejemplos (combinaciones que tenemos/combinaciones totales)
  sumfreq = 0
  for w in freqcomb:
    sumfreq += w/X_train.shape[0]#acumulamos frecuencias para cada combinacion
  sumfreq = sumfreq / contador_combinaciones
  print("avgIR -> "+ str(sumfreq))#avgIR=media del grado de imbalance de todas las etiquetas, se calcula la frecuencia de aparicion de cada combinación
  #y a se divide entre el total de apariciones(instancias totales). Una vez tenemos la frecuencia de cada una se suman y se divide entre la combinaciones posibles
  
  print("rDep -> ")#para calcular esta variable primero sería necesario calcular el número de etiquetas dependientes, usando chi cuadrado (debe superar 6.635 ya que es 99%)
  #y se realiza el promedio del numero total de pares etiqueta
  
  print("\n")






scene:train - exists, not redownloading
number of instances -> 1211
number of attributes -> 294
number of labels -> 6
cardinality -> 1.0619322873658135
density -> 0.1769887145609689
diversity -> 0.21875
avgIR -> 0.07148755455939601
rDep -> 


Corel5k:train - exists, not redownloading
number of instances -> 4500
number of attributes -> 499
number of labels -> 374
cardinality -> 3.5215555555555556
density -> 0.00941592394533571
diversity -> 7.601643391127833e-110
avgIR -> 0.00034195631528966035
rDep -> 


bibtex:train - exists, not redownloading
number of instances -> 4880
number of attributes -> 1836
number of labels -> 159
cardinality -> 2.380327868852459
density -> 0.014970615527373957
diversity -> 2.816281483965306e-45
avgIR -> 0.0004860082206184549
rDep -> 


enron:train - exists, not redownloading
number of instances -> 1123
number of attributes -> 1001
number of labels -> 53
cardinality -> 3.3864648263579697
density -> 0.06389556276147113
diversity -> 6.050715484207103e-14
avgIR -

In [ ]:
#se copia la celda para separar la lista de los dataset

In [ ]:

list = ['rcv1subset4', 'genbase', 'birds', 'emotions','rcv1subset2', 'mediamill', 'medical', 'yeast']

for i in list:
  X_train, y_train, feature_names, label_names = load_dataset(i, 'train')
  
  print(("number of instances -> " + str(X_train.shape[0])))#instancias
  
  print(("number of attributes -> " + str(X_train.shape[1])))#atributos
  
  print(("number of labels -> " + str(y_train.shape[1])))#etiquetas
  
  print("cardinality -> "+ str(y_train.sum()/X_train.shape[0]))#cardinalidad=promedio de etiquetas asociadas a cada instancia (total de 1/instancias)
  
  print("density -> "+ str((y_train.sum()/X_train.shape[0])/y_train.shape[1]))#densidad=cardinalidad entre el número total de etiquetas
  
  #codigo para sacar todas las combinaciones que tenemos y la frecuencia de cada combinacion
  uniquecomb = [y_train.getrow(0),]
  freqcomb = [1,]
  contador_combinaciones = 1
  for w in range(y_train.shape[0]):
    combunica = True
    for z in range(len(uniquecomb)):
      if (y_train.getrow(w).toarray() == uniquecomb[z]).all():
        combunica = False
        freqcomb[z] += 1
        break
    if combunica:
      uniquecomb.append(y_train.getrow(w))
      freqcomb.append(1)
      contador_combinaciones += 1 
  #for w in uniquecomb:
    #print(w.toarray()) 
  #print(contador)
  n_posibles_salidas= 2**y_train.shape[1]
  print("diversity -> "+ str(contador_combinaciones/n_posibles_salidas))#diversidad=medida Distinct(número de combinaciones de distintas etiquetas presentes en un conjunto de datos) 
  
  #normalizada por el número total de ejemplos (combinaciones que tenemos/combinaciones totales)
  sumfreq = 0
  for w in freqcomb:
    sumfreq += w/X_train.shape[0]#acumulamos frecuencias para cada combinacion
  sumfreq = sumfreq / contador_combinaciones
  print("avgIR -> "+ str(sumfreq))#avgIR=media del grado de imbalance de todas las etiquetas, se calcula la frecuencia de aparicion de cada combinación
  #y a se divide entre el total de apariciones(instancias totales). Una vez tenemos la frecuencia de cada una se suman y se divide entre la combinaciones posibles
  
  print("rDep -> ")#para calcular esta variable primero sería necesario calcular el número de etiquetas dependientes, usando chi cuadrado (debe superar 6.635 ya que es 99%)
  #y se realiza el promedio del numero total de pares etiqueta
  
  print("\n")






rcv1subset4:train - does not exists downloading
Downloaded rcv1subset4-train
number of instances -> 3000
number of attributes -> 47229
number of labels -> 101
cardinality -> 2.9103333333333334
density -> 0.028815181518151815
diversity -> 2.007651003787475e-28
avgIR -> 0.001965291421087085
rDep -> 


genbase:train - does not exists downloading
Downloaded genbase-train
number of instances -> 463
number of attributes -> 1186
number of labels -> 27
cardinality -> 1.2613390928725703
density -> 0.04671626269898409
diversity -> 2.1606683731079102e-07
avgIR -> 0.0345572354211663
rDep -> 


birds:train - does not exists downloading
Downloaded birds-train
number of instances -> 322
number of attributes -> 260
number of labels -> 19
cardinality -> 1.0590062111801242
density -> 0.05573716900948022
diversity -> 0.0001697540283203125
avgIR -> 0.011270849326540554
rDep -> 


emotions:train - does not exists downloading
Downloaded emotions-train
number of instances -> 391
number of attributes -> 72
nu

#**Apartado 4**


In [ ]:
!pip install wheel
!pip install scikit-multilearn #Successfully installed scikit-multilearn-0.2.0
!pip install liac-arff
!pip install scipy
!pip install requests
!pip install scikit-learn

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [ ]:
#cargo el dataset de banderas

In [ ]:
%%bash
wget https://sourceforge.net/projects/mulan/files/datasets/flags.zip
unzip flags.zip

In [ ]:
#importo librerías necesarias

In [ ]:
from skmultilearn.dataset import load_from_arff
from sklearn.preprocessing import OneHotEncoder

In [ ]:
import pandas as pd
import arff, numpy as np
from scipy import sparse

In [ ]:
#cargo el dataset

In [ ]:

> Si descar








> Si> Si> Si dataset = arff.load(open('flags.arff'))
data = np.array(dataset['data'])

In [ ]:
#muestro el dataset para comprobar que se ha cargado correctamente

In [ ]:
data

array([['5', '1', '648.0', ..., '1', '1', '0'],
       ['3', '1', '29.0', ..., '0', '1', '0'],
       ['4', '1', '2388.0', ..., '1', '0', '0'],
       ...,
       ['4', '2', '905.0', ..., '0', '0', '1'],
       ['4', '2', '753.0', ..., '0', '1', '1'],
       ['4', '2', '391.0', ..., '1', '1', '0']], dtype='<U32')

In [ ]:
#divido el dataset en entradas y salidas

In [ ]:
X_train = data[:,1:]
y_train = data[:,:1]

In [ ]:
#muestro si se ha dividido correctamente

In [ ]:
print(y_train[:10])

[['5']
 ['3']
 ['4']
 ['6']
 ['3']
 ['4']
 ['1']
 ['1']
 ['2']
 ['2']]


In [ ]:
#aplico onehotencoder para transformar las etiquetas

In [ ]:
encoder = OneHotEncoder(sparse=False, handle_unknown='error')
y_train = encoder.fit_transform(y_train)

In [ ]:
#muestro si se ha dividido correctamente

In [ ]:
print(y_train[:10])

[[0. 0. 0. 0. 1. 0.]
 [0. 0. 1. 0. 0. 0.]
 [0. 0. 0. 1. 0. 0.]
 [0. 0. 0. 0. 0. 1.]
 [0. 0. 1. 0. 0. 0.]
 [0. 0. 0. 1. 0. 0.]
 [1. 0. 0. 0. 0. 0.]
 [1. 0. 0. 0. 0. 0.]
 [0. 1. 0. 0. 0. 0.]
 [0. 1. 0. 0. 0. 0.]]


In [ ]:
#transformo numpy array a lil_matrix

In [ ]:
X_train = np.array(X_train, dtype=float)
y_train = np.array(y_train, dtype=float)
X_train = sparse.csr_matrix(X_train)   # Here's the initialization of the sparse matrix.
y_train = sparse.csr_matrix(y_train)

In [ ]:
#llamo el código anterior para sacar los estadísticos

In [ ]:
print(("number of instances -> " + str(X_train.shape[0])))#instancias

print(("number of attributes -> " + str(X_train.shape[1])))#atributos

print(("number of labels -> " + str(y_train.shape[1])))#etiquetas

print("cardinality -> "+ str(y_train.sum()/X_train.shape[0]))#cardinalidad=promedio de etiquetas asociadas a cada instancia (total de 1/instancias)

print("density -> "+ str((y_train.sum()/X_train.shape[0])/y_train.shape[1]))#densidad=cardinalidad entre el número total de etiquetas

#codigo para sacar todas las combinaciones que tenemos y la frecuencia de cada combinacion
uniquecomb = [y_train.getrow(0),]
freqcomb = [1,]
contador_combinaciones = 1
for w in range(y_train.shape[0]):
  combunica = True
  for z in range(len(uniquecomb)):
    if (y_train.getrow(w).toarray() == uniquecomb[z]).all():
      combunica = False
      freqcomb[z] += 1
      break
  if combunica:
    uniquecomb.append(y_train.getrow(w))
    freqcomb.append(1)
    contador_combinaciones += 1 
#for w in uniquecomb:
  #print(w.toarray()) 
#print(contador)
n_posibles_salidas= 2**y_train.shape[1]
print("diversity -> "+ str(contador_combinaciones/n_posibles_salidas))#diversidad=medida Distinct(número de combinaciones de distintas etiquetas presentes en un conjunto de datos) 

#normalizada por el número total de ejemplos (combinaciones que tenemos/combinaciones totales)
sumfreq = 0
for w in freqcomb:
  sumfreq += w/X_train.shape[0]#acumulamos frecuencias para cada combinacion
sumfreq = sumfreq / contador_combinaciones
print("avgIR -> "+ str(sumfreq))#avgIR=media del grado de imbalance de todas las etiquetas, se calcula la frecuencia de aparicion de cada combinación
#y a se divide entre el total de apariciones(instancias totales). Una vez tenemos la frecuencia de cada una se suman y se divide entre la combinaciones posibles

print("rDep -> ")#para calcular esta variable primero sería necesario calcular el número de etiquetas dependientes, usando chi cuadrado (debe superar 6.635 ya que es 99%)
#y se realiza el promedio del numero total de pares etiqueta

print("\n")

number of instances -> 194
number of attributes -> 25
number of labels -> 6
cardinality -> 1.0
density -> 0.16666666666666666
diversity -> 0.09375
avgIR -> 0.16752577319587625
rDep -> 




In [ ]:
#cargo el dataset de ranas

In [ ]:
%%bash
wget https://archive.ics.uci.edu/ml/machine-learning-databases/00528/dataset.csv

--2023-01-16 11:37:21--  https://archive.ics.uci.edu/ml/machine-learning-databases/00528/dataset.csv
Resolving archive.ics.uci.edu (archive.ics.uci.edu)... 128.195.10.252
Connecting to archive.ics.uci.edu (archive.ics.uci.edu)|128.195.10.252|:443... connected.
HTTP request sent, awaiting response... 200 OK
Length: 10846 (11K) [application/x-httpd-php]
Saving to: ‘dataset.csv.1’

     0K ..........                                            100% 73.2M=0s

2023-01-16 11:37:21 (73.2 MB/s) - ‘dataset.csv.1’ saved [10846/10846]



In [ ]:
#leo el csv

In [ ]:
data = pd.read_csv("dataset.csv",sep=';')

In [ ]:
data

Integer Categorical Numerical Numerical.1 Categorical.1 Categorical.2  \
0        ID    Motorway        SR          NR            TR            VR   
1         1          A1       600           1             1             4   
2         2          A1       700           1             5             1   
3         3          A1       200           1             5             1   
4         4          A1       300           1             5             0   
..      ...         ...       ...         ...           ...           ...   
185     185         S52      2300           1            12             3   
186     186         S52       300           1            14             2   
187     187         S52       500           1             1             4   
188     188         S52       300           1            12             3   
189     189         S52       300           1            12             3   

    Categorical.3 Categorical.4 Categorical.5 Categorical.6  ... Ordinal.1  \
0            SUR1          SUR2          SUR3            UR  ...        BR   
1               6             2            10             0  ...         0   
2              10             6            10             3  ...         1   
3              10             6            10             3  ...         1   
4               6            10             2             3  ...         0   
..            ...           ...           ...           ...  ...       ...   
185             2             2             1             0  ...         1   
186             7            10             2             0  ...         5   
187             1            10             2             0  ...         5   
188             2             1             6             0  ...         0   
189             2             6            10             0  ...         1   

    Categorical.8 Categorical.9      Label 1      Label 2      Label 3  \
0              MR            CR  Green frogs  Brown frogs  Common toad   
1               0             1            0            0            0   
2               0             1            0            1            1   
3               0             1            0            1            1   
4               0             1            0            0            1   
..            ...           ...          ...          ...          ...   
185             0             1            0            1            0   
186             0             1            1            1            1   
187             0             1            1            1            1   
188             0             1            0            1            1   
189             0             1            0            1            1   

               Label 4    Label 5      Label 6             Label 7  
0    Fire-bellied toad  Tree frog  Common newt  Great crested newt  
1                    0          0            0                   0  
2                    0          0            1                   0  
3                    0          0            1                   0  
4                    0          0            0                   0  
..                 ...        ...          ...                 ...  
185                  0          0            0                   0  
186                  1          0            1                   0  
187                  1          0            1                   0  
188                  0          0            0                   0  
189                  0          0            0                   0  

[190 rows x 23 columns]

In [ ]:
#elimino columnas que no me interesan

In [ ]:
data = data.drop(data.columns[[1]],axis = 1)

In [ ]:
#transformo a numpy

In [ ]:
data = data.to_numpy()

In [ ]:
#elimino filas con información

In [ ]:
data = np.delete(data, 0, 0)

In [ ]:
#transformo la matriz numpy a int

In [ ]:
data = np.int32(data)

In [ ]:
#muestro los datos

In [ ]:
print(data)

[[  1 600   1 ...   0   0   0]
 [  2 700   1 ...   0   1   0]
 [  3 200   1 ...   0   1   0]
 ...
 [187 500   1 ...   0   1   0]
 [188 300   1 ...   0   0   0]
 [189 300   1 ...   0   0   0]]


In [ ]:
#los divido en entradas y salidas

In [ ]:
X_train = data[:,:-7]
y_train = data[:,-7:]

In [ ]:
print(y_train[:10])

[[0 0 0 0 0 0 0]
 [0 1 1 0 0 1 0]
 [0 1 1 0 0 1 0]
 [0 0 1 0 0 0 0]
 [0 1 1 1 0 1 1]
 [0 0 0 0 0 0 0]
 [0 0 1 0 0 0 0]
 [1 1 0 1 0 0 0]
 [0 1 1 0 0 0 0]
 [0 1 1 0 0 0 0]]


In [ ]:
#finalmente transformo numpy en lil_matrix

In [ ]:
X_train = np.array(X_train, dtype=float)
y_train = np.array(y_train, dtype=float)
X_train = sparse.csr_matrix(X_train)   # Here's the initialization of the sparse matrix.
y_train = sparse.csr_matrix(y_train)

In [ ]:
#llamo el código anterior para sacar los estadísticos

In [ ]:
print(("number of instances -> " + str(X_train.shape[0])))#instancias

print(("number of attributes -> " + str(X_train.shape[1])))#atributos

print(("number of labels -> " + str(y_train.shape[1])))#etiquetas

print("cardinality -> "+ str(y_train.sum()/X_train.shape[0]))#cardinalidad=promedio de etiquetas asociadas a cada instancia (total de 1/instancias)

print("density -> "+ str((y_train.sum()/X_train.shape[0])/y_train.shape[1]))#densidad=cardinalidad entre el número total de etiquetas

#codigo para sacar todas las combinaciones que tenemos y la frecuencia de cada combinacion
uniquecomb = [y_train.getrow(0),]
freqcomb = [1,]
contador_combinaciones = 1
for w in range(y_train.shape[0]):
  combunica = True
  for z in range(len(uniquecomb)):
    if (y_train.getrow(w).toarray() == uniquecomb[z]).all():
      combunica = False
      freqcomb[z] += 1
      break
  if combunica:
    uniquecomb.append(y_train.getrow(w))
    freqcomb.append(1)
    contador_combinaciones += 1 
#for w in uniquecomb:
  #print(w.toarray()) 
#print(contador)
n_posibles_salidas= 2**y_train.shape[1]
print("diversity -> "+ str(contador_combinaciones/n_posibles_salidas))#diversidad=medida Distinct(número de combinaciones de distintas etiquetas presentes en un conjunto de datos) 

#normalizada por el número total de ejemplos (combinaciones que tenemos/combinaciones totales)
sumfreq = 0
for w in freqcomb:
  sumfreq += w/X_train.shape[0]#acumulamos frecuencias para cada combinacion
sumfreq = sumfreq / contador_combinaciones
print("avgIR -> "+ str(sumfreq))#avgIR=media del grado de imbalance de todas las etiquetas, se calcula la frecuencia de aparicion de cada combinación
#y a se divide entre el total de apariciones(instancias totales). Una vez tenemos la frecuencia de cada una se suman y se divide entre la combinaciones posibles

print("rDep -> ")#para calcular esta variable primero sería necesario calcular el número de etiquetas dependientes, usando chi cuadrado (debe superar 6.635 ya que es 99%)
#y se realiza el promedio del numero total de pares etiqueta

print("\n")

number of instances -> 189
number of attributes -> 15
number of labels -> 7
cardinality -> 3.111111111111111
density -> 0.4444444444444445
diversity -> 0.3359375
avgIR -> 0.023378860588162926
rDep -> 




#**Apartado 5**

In [ ]:
!pip install wheel
!pip install scikit-multilearn #Successfully installed scikit-multilearn-0.2.0
!pip install liac-arff
!pip install scipy
!pip install requests
!pip install scikit-learn

In [ ]:
from skmultilearn.dataset import load_dataset

In [ ]:
#ML transformación

In [ ]:
from skmultilearn.problem_transform import BinaryRelevance
from skmultilearn.problem_transform import ClassifierChain
from skmultilearn.problem_transform import LabelPowerset
from sklearn.svm import SVC

from sklearn.metrics import confusion_matrix, accuracy_score, precision_score, recall_score, f1_score

In [ ]:
# initialize Binary Relevance multi-label classifier
# with an SVM classifier
# SVM in scikit only supports the X matrix in sparse representation

X_train, y_train, feature_names, label_names = load_dataset('scene', 'train')
X_test, y_test, feature_names, label_names = load_dataset('scene', 'test')

classifier = BinaryRelevance(
    classifier = SVC(),
    require_dense = [False, True]
)

# train
classifier.fit(X_train, y_train)

# predict
predictions = classifier.predict(X_test)
accuracy_score(y_test, predictions)


scene:train - exists, not redownloading
scene:test - exists, not redownloading


0.5869565217391305

In [ ]:

# initialize Classifier Chain multi-label classifier
# with an SVM classifier
# SVM in scikit only supports the X matrix in sparse representation

X_train, y_train, feature_names, label_names = load_dataset('scene', 'train')
X_test, y_test, feature_names, label_names = load_dataset('scene', 'test')

classifier = ClassifierChain(
    classifier = SVC(),
    require_dense = [False, True]
)

# train
classifier.fit(X_train, y_train)

# predict
predictions = classifier.predict(X_test)
accuracy_score(y_test, predictions)

scene:train - exists, not redownloading
scene:test - exists, not redownloading


0.6780936454849499

In [ ]:
# initialize LabelPowerset multi-label classifier with a SVM classifier

X_train, y_train, feature_names, label_names = load_dataset('scene', 'train')
X_test, y_test, feature_names, label_names = load_dataset('scene', 'test')

classifier = ClassifierChain(
    classifier = SVC(),
    require_dense = [False, True]
)

# train
classifier.fit(X_train, y_train)

# predict
predictions = classifier.predict(X_test)
accuracy_score(y_test, predictions)

scene:train - exists, not redownloading
scene:test - exists, not redownloading


0.6780936454849499

In [ ]:
#ML adaptación

In [ ]:
from skmultilearn.adapt import MLkNN
from skmultilearn.adapt import BRkNNaClassifier
from skmultilearn.adapt import MLTSVM

In [ ]:
#MLkNN

X_train, y_train, feature_names, label_names = load_dataset('scene', 'train')
X_test, y_test, feature_names, label_names = load_dataset('scene', 'test')

classifier = MLkNN(k=3)

# train
#classifier.fit(X_train, y_train)#actualmente no funciona esta funcion

# predict
#predictions = classifier.predict(X_test)
#accuracy_score(y_test, predictions)

scene:train - exists, not redownloading
scene:test - exists, not redownloading


In [ ]:
X_train, y_train, feature_names, label_names = load_dataset('scene', 'train')
X_test, y_test, feature_names, label_names = load_dataset('scene', 'test')

#BRkNNaClassifier

classifier = BRkNNaClassifier(k=3)

# train
#classifier.fit(X_train, y_train)#actualmente no funciona esta funcion

# predict
#predictions = classifier.predict(X_test)
#accuracy_score(y_test, predictions)

scene:train - exists, not redownloading
scene:test - exists, not redownloading


In [ ]:
#MLTSVM

X_train, y_train, feature_names, label_names = load_dataset('scene', 'train')
X_test, y_test, feature_names, label_names = load_dataset('scene', 'test')

classifier = MLTSVM(c_k = 2**-1)

# train
classifier.fit(X_train, y_train)

# predict
predictions = classifier.predict(X_test)
accuracy_score(y_test, predictions)

scene:train - exists, not redownloading
scene:test - exists, not redownloading


0.22408026755852842

#**Apartado 6**

In [ ]:
#KFold() y cross_validate(), cross_val_score() y make_scorer()
#
#KFold()->Proporciona indices de train y test para dividir el conjunto de entrenamiento y test, es decir, divide el conjunto en k partes. Siendo k-1 partes el conjunto de entrenamiento en cada caso
#
#cross_validate()->evalúa las métricas aplicando validación cruzada
#
#cross_val_score()->evalúa la puntuación aplicando validación cruzada
#
#make_scorer()->crea una puntuación a partir de una métrica de rendimiento o una función de pérdida

In [ ]:
#Para aplicar validación cruzada en el contexto ML, se realizaría con k_fold haciendo estratificación iterativa, obteniendo así los resultados para cada fold

In [ ]:
!pip install wheel
!pip install scikit-multilearn #Successfully installed scikit-multilearn-0.2.0
!pip install liac-arff
!pip install scipy
!pip install requests
!pip install scikit-learn

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [ ]:
from skmultilearn.model_selection import IterativeStratification
from skmultilearn.dataset import load_dataset
from skmultilearn.problem_transform import BinaryRelevance
from sklearn.svm import SVC

In [ ]:
X_train, y_train, feature_names, label_names = load_dataset('scene', 'train')
X_test, y_test, feature_names, label_names = load_dataset('scene', 'test')#esto se usaría si se quiere validar despues de la validación cruzada, ya que sino usaríamos datos de test en el entrenamiento

classifier = BinaryRelevance(
    classifier = SVC(),
    require_dense = [False, True]
)
sol = []

k_fold = IterativeStratification(n_splits=2, order=1)
for train, test in k_fold.split(X_train, y_train):
    classifier.fit(X_train[train], y_train[train])
    result = classifier.predict(X_train[test])
    sol.append(accuracy_score(y_train[test], result))
print(sol)

scene:train - exists, not redownloading
scene:test - exists, not redownloading
[0.5247524752475248, 0.5586776859504132]


#**Apartado 7**

In [ ]:
#Existen dos tipos de Métricas evaluación, pudiendo ser Basadas en instancias o Basadas en etiquetas.
#
#Las basadas en etiquetas son:
#
#Accuracy-> la proporción de predichas correctamente respecto al total
#
#Recall-> la fracción de etiquetas positivas reales que se consiguieron predecir (en aprendizaje con imbalance, se emplea recall para medir la cobertura de la clase minoritaria)
#
#F1_score-> en una única medida atiende a las preocupaciones tanto de precision como de recall
#
#Las basadas en instancias son:
#
#Hamming loss-> Evalúa cuántas veces un par instancia-etiqueta se clasifica mal. Operación XOR
#
#Precision-> calcular la media de la precisión de los clasificadores binarios
#
#Cabe destacar el uso del area bajo la curva roc o la matriz de confusion para cada clase o para el conjunto de etiquetas, ya que se puede conseguir información relevante a partir de los parámetros anteriores como F1_score o Accuracy o recall.

#**Apartado 8**

In [ ]:
#cl-cv.py

In [ ]:
!pip install wheel
!pip install scikit-multilearn #Successfully installed scikit-multilearn-0.2.0
!pip install liac-arff
!pip install scipy
!pip install requests
!pip install scikit-learn

Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/
Looking in indexes: https://pypi.org/simple, https://us-python.pkg.dev/colab-wheels/public/simple/


In [ ]:
#uso los imports necesarios

In [ ]:
from skmultilearn.model_selection import IterativeStratification
from skmultilearn.dataset import load_dataset
from skmultilearn.problem_transform import BinaryRelevance
from skmultilearn.problem_transform import ClassifierChain
from skmultilearn.adapt import MLTSVM
from sklearn.svm import SVC
from sklearn.metrics import classification_report
import warnings
warnings.filterwarnings("ignore")

In [ ]:
#cargo el dataset, realizo kfold de 3 split para cada clasificador, los clasificadores son BinaryRelevance, ClassifierChain y MLTSVM

In [ ]:
list = ['tmc2007_500','yeast', 'scene', 'birds','emotions']#lista con los 5 dataset

for i in list:
  X_train, y_train, feature_names, label_names = load_dataset(i, 'train')
  X_test, y_test, feature_names, label_names = load_dataset(i, 'test')#esto se usaría si se quiere validar despues de la validación cruzada, ya que sino usaríamos datos de test en el entrenamiento


  #clasificador BinaryRelevance
  classifier = BinaryRelevance(
    classifier = SVC(),
    require_dense = [False, True]
  )
  sol = []

  k_fold = IterativeStratification(n_splits=3)
  for train, test in k_fold.split(X_train, y_train):
      classifier.fit(X_train[train], y_train[train])
      result = classifier.predict(X_train[test])
      sol.append(classification_report(y_train[test], result))
  print("Resultados para BinaryRelevance en la bd ", i)
  for w in sol:
    print(w)

  #clasificador ClassifierChain
  classifier = ClassifierChain(
    classifier = SVC(),
    require_dense = [False, True]
)
  sol = []

  k_fold = IterativeStratification(n_splits=3)
  for train, test in k_fold.split(X_train, y_train):
      classifier.fit(X_train[train], y_train[train])
      result = classifier.predict(X_train[test])
      sol.append(classification_report(y_train[test], result))
  print("Resultados para ClassifierChain en la bd", i)
  for w in sol:
    print(w)

  #clasificador MLTSVM
  classifier = MLTSVM(c_k = 2**-1)
  sol = []

  k_fold = IterativeStratification(n_splits=3)
  for train, test in k_fold.split(X_train, y_train):
      classifier.fit(X_train[train], y_train[train])
      result = classifier.predict(X_train[test])
      sol.append(classification_report(y_train[test], result))
  print("Resultados para MLTSVM en la bd ", i)
  for w in sol:
    print(w)


tmc2007_500:train - exists, not redownloading
tmc2007_500:test - exists, not redownloading
Resultados para BinaryRelevance en la bd  tmc2007_500
              precision    recall  f1-score   support

           0       0.76      0.61      0.68       478
           1       0.72      0.88      0.79      4292
           2       0.76      0.52      0.62       131
           3       0.88      0.15      0.26       144
           4       0.84      0.72      0.77      1021
           5       0.75      0.68      0.71      2021
           6       0.55      0.01      0.02       553
           7       0.82      0.43      0.57       737
           8       0.80      0.03      0.06       137
           9       0.81      0.13      0.22       369
          10       0.96      0.19      0.32       118
          11       0.86      0.64      0.73      1119
          12       0.76      0.33      0.46       720
          13       0.84      0.54      0.66       421
          14       0.00      0.00      0.00 

#**Apartado 9**

*   ¿Qué método(s) parece(n) comportarse mejor globalmente? Entre otros aspectos, analiza el efecto que tiene en el rendimiento de los métodos de transformación la elección que hagamos del clasificador base.

> Respecto a los métodos que se han comprobado en los apartados anteriores, podemos destacar que los que se suelen comportar mejor globalmente son los ML transformación, en concreto quizás un poco más ClassifierChain, aunque lógicamente dependerá del dataset que se este utilizando.
Lógicamente, la elección que hagamos del clasificador tendrá relevancia sobre cualquiera de los métodos de transformación, ya que si dividieramos el problema como si de una sola etiqueta se tratasé, el rendimiento incidirá en dicho clasificador. Por lo tanto si un dataset se divide en diferentes dataset con una sola etiqueta, aquel clasificador que funcione mejor para dicho dataset, mejorar el resultado del mejor clasificador. Dependiendo del dataset, será mejor usar SVM, árboles de decisión o knn, entre otros.


*   ¿Cuál es el tiempo de ejecución de cada método? ¿Existen diferencias  destacables?

> Si comparamos los tiempos entre ML de transformación y ML de adaptación, llegamos a la conclusión que dependerá del dataset. De tal forma que, si por ejemplo un dataset tiene pocas labels como puede ser 6, los clasificadores de transformación serán más rápidos que los de adaptación, si es el caso contrario y tenemos un dataset con 120 etiquetas, los clasificadores de transformación serán muy lentos, ya que para cada etiqueta se esta generando un clasificador propio. Por lo tanto, los tiempos dependerán principalmente del dataset que se utilice y de la estructura del mismo.

*   ¿Observas diferencias reseñables en cuanto a los valores de cada métrica?

> Si descartamos los valores dados por los dataset que tienen pocas instancias, y por tanto no se realizan de forma precisa los split generando resultados muy imprecisos. Podemos observar que los valores de precision, recall y f1-score nos dan información relevante sobre cada dataset con su respectivo clasificador.
De tal forma que, 

*   Analiza si observas alguna relación entre los resultados obtenidos para las métricas de evaluación y las características propias del conjunto ML: número de instancias, número de etiquetas, densidad, cardinalidad...? Etc. Escribe un breve informe sobre si los estadísticos calculados en el entregable1 dan información del rendimiento que estás observando. ¿Se pueden extraer conclusiones al respecto? Puedes buscar información en referencias bibliográficas previas.

> Si nos fijamos en los estadisticos obtenidos, los cuales son, el número de instancias, número de atributos, número de etiquetas, cardinalidad, densidad, diversidad y el avgIr (se obvia el rDep ya que no se pudo obtener), para cada dataset obtenido y lo comparamos con los resultados obtenidos, podemos extraer la siguiente información, antes de ello, se muestran los estadisticos de los dataset usados para los distintos clasificadores multietiqueta (BinaryRelevance, ClassifierChain y MLTSVM) para recordarlos:

>**birds**: <br>
number of instances -> 322<br>
number of attributes -> 260<br>
number of labels -> 19<br>
cardinality -> 1.059<br>
density -> 0.055<br>
diversity -> 0.0001<br>
avgIR -> 0.011<br>

>**emotions**:<br>
number of instances -> 391<br>
number of attributes -> 72<br>
number of labels -> 6<br>
cardinality -> 1.813<br>
density -> 0.302<br>
diversity -> 0.406<br>
avgIR -> 0.038<br>

>**yeast**:<br>
number of instances -> 1500<br>
number of attributes -> 103<br>
number of labels -> 14<br>
cardinality -> 4.228<br>
density -> 0.302<br>
diversity -> 0.0100<br>
avgIR -> 0.006<br>

>**tmc2007_500**:<br>
number of instances -> 21519<br>
number of attributes -> 500<br>
number of labels -> 22<br>
cardinality -> 2.225<br>
density -> 0.101<br>
diversity -> 0.0002<br>
avgIR -> 0.0008<br>

>**scene**:<br>
number of instances -> 1211<br>
number of attributes -> 294<br>
number of labels -> 6<br>
cardinality -> 1.061<br>
density -> 0.176<br>
diversity -> 0.218<br>
avgIR -> 0.071<br>

> Cabe destacar que los se usaron diferentes tipo de clasificadores, siendo de transformación como BinaryRelevance y ClassifierChain, o adaptación como MLTSVM. Para todos ellos se aplica un k-fold de tres.

La información que se ha podido obtener tras realizar este experimento es la siguiente:


> - En general, a mayor cantidad de instancias y atributos, el tiempo de computo aumenta en mayor medida, siendo el dataset tmc2007_500, el que mayor tiempo de computo demoró para los tres clasificadores.


> - Aquellos dataset con un número de etiquetas mayor(birds y tmc2007_500)aumentan su tiempo de cómputo para los clasificadores de tranformación ya que en el caso de BinaryRelevance, se utiliza en este caso SVC y para cada par de etiquetas se esta generando un clasificador, por tanto mientras mayor sea el número de etiquetas mayor será el tiempo de computo,frente a algoritmos adaptados para este tipo de problemas multietiqueta.

> - Aquellos dataset con un avgIR mayor como scene y emotions, presentan un invalanceo de etiquetas, y por tanto, si nos fijamos en el accuracy de estos los valores suelen ser bastante dispares entre sí. Este se debe a que a la hora de entrenar, no todas las etiquetas tienen cantidades de instancias similares para entrenar.

> - Los dataset de birds y emotions, dan mejores resultados con MLTSVM que con los clasificadores de transformación, es probable que se deba al dataset en sí y no a los estadísticos.

> - Respecto a cual da mejores resultados, este es el dataset yeast, ya que es el dataset con mayor número de instancias y por tanto, la cardinalidad de dicho dataset es muy alta. Esto influye a que a la hora de entrenar cada instancia tiene como media unas cuatro etiquetas, permitiendo así que algoritmos de adaptación puedan aprender las relaciones entre etiquetas. Aun así, también funciona bien para los algoritmos de transformación por la gran cantidad de instancias.


